Allow plural outputs to be cached

lptr opened the pull request:
When `@OutputFiles` or `@OutputDirectories` is used with an `Iterable` (instead of a `Map`), we previously disabled caching without much explanation. The original reason for this was that we expected a difference in cardinality of outputs between the task producing the output and the one consuming it. However, if we represent each entry in the collection as an indexed property, then unmatched cardinality will produce a different cache key anyway. Thus there's no need for this limitation.

lptr commented:
Branch build: https://builds.gradle.org/viewQueued.html?itemId=15290212

wolfs commented on subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/TaskCacheabilityReasonIntegrationTest.groovy:
```diff
@@ -133,20 +133,22 @@ class TaskCacheabilityReasonIntegrationTest extends AbstractIntegrationSpec impl
             task cacheable(type: PluralOutputs) {
                 cachingEnabled = false
-                disabledReason = "Declares multiple output files for the single output property 'outputFiles' via `@OutputFiles`, `@OutputDirectories` or `TaskOutputs.files()`"
+                disabledReason = "Declares complex output for the output property 'outputFiles'"
```
That will probably require an update in build scans as well. Currently, the reason is "Multiple outputs declared": 
![image](https://user-images.githubusercontent.com/423186/44711330-67f5e600-aaae-11e8-80ea-e5938c329691.png)

I didn't find any help text for it, I guess we would need to add a section to the user guide or the build cache guide explaining what kind of outputs we support for caching and link to it. WDYT?

wolfs replied:
I also think that `complex` is very vague. Can we be more specific in what is allowed? Something like "Only files, directories and unfiltered file trees are allowed"?

wolfs commented on subprojects/docs/src/docs/release/notes.md:
```diff
@@ -58,6 +58,14 @@ In this way, stale execution data, which would cause non-repeatable task outputs
 Since Gradle now takes care of removing the execution data, the `JacocoPluginExtension.append` property has been deprecated.
 The JaCoCo agent is always configured with `append = true`, so it can be used when running tests in parallel. 
 
+### Plural task output properties don't disable caching anymore
```
I think we also need to add something/change the user guide and adjust the Javadocs on `OutputFiles` and `OutputDirectories`.

lptr commented:
I've added a new non-caching reason `NON_CACHEABLE_TREE_OUTPUT`. 

I also disabled caching if any tree appears in the output for the following reasons:

* it's easier to explain to users,
* it's unlikely that people would want to use an unfiltered `DirectoryFileTree` (which is an internal type anyway) as an output,
* we don't support `DirectoryFileTree` as an `@OutputDirectory` either (we actually throw an exception).

I've updated our horrible user guide chapter about task inputs with the changes.

lptr commented:
Branch build: https://builds.gradle.org/viewQueued.html?itemId=15343943

lptr commented:
This will need to be mentioned in https://github.com/gradle/dotcom/issues/1399, too.

lptr commented:
Branch build: https://builds.gradle.org/viewQueued.html?itemId=15349354

lptr commented:
Based on @facewindu's feedback I removed the old enum value.
